app/models/manageiq/providers/hawkular/middleware_manager.rb

```diff
@@ -521,6 +562,54 @@ def run_operation(parameters, operation_name = nil)
       end
     end
 
+    EmitNotificationArgs = Struct.new(:type, :operation_name, :operation_args,
+                                      :target_resource, :entity_klass, :detailed_message)
+    private_constant :EmitNotificationArgs
+
+    def emit_middleware_notification(notification_args)
+      mw_entity = notification_args.entity_klass.find_by(:ems_ref => ems_ref) unless notification_args.entity_klass == MiddlewareServer
+      mw_server = if mw_entity.nil?
+                    MiddlewareServer.find_by(:ems_ref => notification_args.target_resource) ||
+                      MiddlewareDomain.find_by(:ems_ref => notification_args.target_resource)
+                  else
+                    MiddlewareServer.find_by(:id => mw_entity.server_id)
+                  end
+
+      return unless mw_server
+
+      Notification.create(
+        :type => notification_args.type, :options => {
+          :op_name   => notification_args.operation_name,
+          :op_arg    => notification_args.operation_args || '',
+          :mw_server => "#{mw_server.name} (#{mw_server.feed})"
+        }
+      )
+
+      unless mw_entity
+        message = _('%{operation} operation for %{server} %{status}') %
+                  {
+                    :operation => notification_args.operation_name,
+                    :server    => mw_server.name,
+                    :status    => notification_args.type == :mw_op_success ? _('succeeded') : _('failed')
+                  }
+        message += ": #{notification_args.detailed_message}" if notification_args.detailed_message
+
+        EmsEvent.add_queue(
+          'add', id,
+          :ems_id          => id,
+          :source          => 'HAWKULAR',
+          :timestamp       => Time.zone.now,
+          :event_type      => '%{entity_type}.%{operation}.%{status}' %
```

Comment:
This could be a method in EmitNotificationArgs itself. The idea behind the args object is to not query the data directly and then decide how to act on it, but to get the final data that we do need, and not introspect on the object itself. So something like that:

```ruby
notification_args.event_type
```